src/rgw/rgw_client_io.h

```diff
@@ -4,83 +4,236 @@
+#include <boost/utility/string_ref.hpp>
 
+#include "include/types.h"
 #include "rgw_common.h"
 
 class RGWClientIO {
-  bool _account;
-
 protected:
-  RGWEnv env;
-
   virtual void init_env(CephContext *cct) = 0;
 
 public:
   virtual ~RGWClientIO() {}
-  RGWClientIO() : _account(false) {}
 
   void init(CephContext *cct);
-  RGWEnv& get_env() { return env; }
+  virtual RGWEnv& get_env() noexcept = 0;
+  virtual int complete_request() = 0;
+}; /* RGWClient IO */
+
+
+class RGWClientIOAccounter {
+public:
+  virtual ~RGWClientIOAccounter() {}
+
+  virtual void set_account(bool enabled) = 0;
+
+  virtual uint64_t get_bytes_sent() const = 0;
+  virtual uint64_t get_bytes_received() const = 0;
+};
+
+
+class RGWRestfulIOEngine : public RGWClientIO {
+  template<typename T> friend class RGWDecoratedRestfulIO;
+
+public:
+  class Exception : public std::exception {
+    int err;
+  public:
+    Exception(const int err)
+      : err(err) {
+    }
 
-  bool account() { return _account; }
-  void set_account(bool _accnt) {
-    _account = _accnt;
+    int value() {
+      return err;
+    }
+  };
+
+  virtual std::size_t send_status(int status, const char *status_name) = 0;
+  virtual std::size_t send_100_continue() = 0;
+
+  /* Send header to client. On success returns number of bytes sent to the direct
+   * client of RadosGW. On failure throws int containing errno. boost::string_ref
+   * is being used because of length it internally carries. */
+  virtual std::size_t send_header(const boost::string_ref& name,
+                                  const boost::string_ref& value) = 0;
+
+  /* Inform a client about a content length. Takes number of bytes supplied in
+   * @len XOR one of the alternative modes for dealing with it passed as @mode.
+   * On success returns number of bytes sent to the direct client of RadosGW.
+   * On failure throws int containing errno.
+   *
+   * CALL ORDER:
+   *  - The method must be called EXACTLY ONE time.
+   *  - The method must be preceeded with a call to send_status().
+   *  - The method must not be called after complete_header(). */
+  virtual std::size_t send_content_length(uint64_t len) = 0;
+
+  virtual std::size_t send_chunked_transfer_encoding() {
+    /* This is a null implementation. We don't send anything here, even the HTTP
+     * header. The intended behaviour should be provided through a decorator or
+     * directly by a given front-end. */
+    return 0;
   }
 
-  virtual int complete_request() = 0; /* XXX signature likely changing */
+  virtual std::size_t complete_header() = 0;
 
-  virtual uint64_t get_bytes_sent() { return 0; }
-  virtual uint64_t get_bytes_received() { return 0; }
-}; /* RGWClient IO */
+  /* Receive body. On success Returns number of bytes sent to the direct
+   * client of RadosGW. On failure throws int containing errno. */
+  virtual std::size_t recv_body(char* buf, std::size_t max) = 0;
+  virtual std::size_t send_body(const char* buf, std::size_t len) = 0;
```

Comment:
`recv_body` and `send_body` are parts of the public interface for doing **restful** IO while`write_data` and `read_data` live in the stream domain and are actually implementation' details. Although some engines are stream-oriented (both headers and body are handled the same way; FastCGI might be seen as an example here), others aren't. This applies especially to filters (accounting, chunking, buffering, reordering). Having to implement a machine state to differentiate headers and body would unnecessarily increase the complexity. Thus, I believe it's better to leverage the fact that the interface' usage pattern is truly restful.